scripts/miniapps.py

```diff
@@ -0,0 +1,128 @@
+from os import system, makedirs
+from os.path import expanduser
+from re import sub
+
+# Finds two factors of `n` that are as close to each other as possible.
+#
+# Note: the second factor is larger or equal to the first factor
+def _sq_factor(n):
+    for i in range(1, int(n ** 0.5) + 1):
+        if n % i == 0:
+            f = (i, n // i)
+    return f
+
+
+def _check_ranks_per_node(rpn):
+    if not (rpn == 1 or rpn == 2):
+        raise ValueError(f"Wrong value rpn = {rpn}!")
+
+
+# parse dlaf_nb_[1|0][1|0][1|0]
+#       dlaf_nb_<hpx queue><non-blocking mechanism><mpi pool>
+#       dlaf
+def _dlaf_flags(lib):
+    if lib == "dlaf":
+        return ""
+
+    q, p, m = tuple(lib[8:])
+    return "{} {} {}".format(
+        "--hpx:queuing=shared-priority" if q == "1" else "",
+        "--polling" if p == "1" else "",
+        "--mpipool --hpx:ini=hpx.max_idle_backoff_time=0" if m == "1" else "",
+    )
+
+
+def _err_msg(lib):
+    return f"No such `lib`: {lib}! Allowed values are : `dlaf|dlaf_nb_[0|1][0|1][0|1]`, `slate` and `dplasma`."
+
+
+# Job preamble
+#
+def init_job_text(run_name, nodes, time_min):
+    job_text = f"""
+#!/bin/bash -l
+#SBATCH --job-name={run_name}_{nodes}
+#SBATCH --time={time_min}
+#SBATCH --nodes={nodes}
+#SBATCH --partition=normal
+#SBATCH --account=csstaff
+#SBATCH --constraint=mc
+#SBATCH --output=output.txt
+#SBATCH --error=error.txt
+
+# Env
+export CRAYPE_LINK_TYPE=dynamic
+export MPICH_MAX_THREAD_SAFETY=multiple
+
+# Debug
+module list &> modules.txt
+printenv > env.txt
+"""
+    return job_text[1:-1]
+
+
+# Create the job directory tree and submit jobs.
+#
+def submit_jobs(run_dir, nodes, job_text):
+    job_path = expanduser(f"{run_dir}/{nodes}")
+    makedirs(job_path, exist_ok=False)
+    job_file = f"{job_path}/job.sh"
+    with open(job_file, "w") as f:
+        f.write(job_text)
+
+    print(f"Submitting : {job_file}")
+    system(f"sbatch --chdir={job_path} {job_file}")
+
+
+# lib: allowed libraries are dlaf|dlaf_nb_[0|1][0|1][0|1]|slate|dplasma
+#
+# where `dlaf_nb_[0|1][0|1][0|1]` is `dlaf_nb_<hpx queue><non-blocking mechanism><mpi pool>`, for example
+#       `dlaf_nb_111` : dlaf with non-blocking MPI(nb), shared priority queue(1), polling (1) and a dedicated core for an MPI pool (1).
+#
+# rpn: ranks per node
+#
+def chol(lib, build_dir, nodes, rpn, m_sz, mb_sz, nruns):
+    _check_ranks_per_node(rpn)
+
+    total_ranks = nodes * rpn
+    cpus_per_rank = 72 // rpn
```

Comment:
I would not hardcode system dependent parameter in a file that (citing the README) usually should not be modified. (Applies to other part of this file.)

Reply:
Maybe we should add config files (json?) for different systems
and let the user specify the system in the gen.py file.